pullrequests/valid_parentheses/step2.go

```diff
@@ -0,0 +1,22 @@
+//lint:file-ignore U1000 Ignore all unused code
+package validparentheses
+
+/*
+ネストと変数名を改善しました。
+*/
+func isValidStep2(s string) bool {
+	var stack []rune
+	closeToOpens := map[rune]rune{')': '(', '}': '{', ']': '['}
+	for _, bracket := range s {
+		openBracket, ok := closeToOpens[bracket]
+		if !ok {
+			stack = append(stack, bracket)
+			continue
+		}
+		if len(stack) == 0 || stack[len(stack)-1] != openBracket {
+			return false
+		}
+		stack = stack[:len(stack)-1]
```

Comment:
あまり重要でない気もする個人的な感想なのですが、 step 1 にあった "comma ok" idiom (と Effective Go で呼んでました) が来なかったのを意外に感じました。

Go だしこんな感じの実装になるかなとイメージしていました:
```suggestion
		if openBracket, ok := closeToOpens[bracket]; ok {
			if len(stack) == 0 || stack[len(stack)-1] != openBracket {
				return false
			}
			stack = stack[:len(stack)-1]
			continue
		}
		stack = append(stack, bracket)
```

Reply:
単純に認知負荷が下がる書き方を選びました。行数的にもほぼ変わらないですし、上の方が認知負荷が下がる分良いかなと思うのですが、下の書き方のメリットは何かあったりしますでしょうか（僕もGoにとても詳しい人間というわけではないので、下の書き方がGoらしいのであれば何か理由があったりするのかなと思ってご存じだったら教えていただきたいなと🙏）。

Reply:
いえ、どちらの書き方も文法的には正しいですし、書き換えることで大きなメリットがあったりなかったりするとは考えていません。
仕事で数年 Go のコードを見ていますが、宣言された ok を使った if 分岐が comma ok idiom で書かれないのが珍しいなと思ったのでコメントした次第なのですが、強いて言うなら、自分が面接官をする場合、上記の idiom で書かれていたら「この人はある程度 Go で開発しているんだろうな」いう印象を持つかなとは思いました。
逆に Go に習熟していることをアピールしている求職者がこのような場合に comma ok idiom を使ってなかったら、ちょっと気になって「これあえてこうしてるんですか？」とか聞くかも知れません。

Reply:
いえ、どちらの書き方も文法的には正しいですし、書き換えることで大きなメリットがあったりなかったりするとは考えていません。
仕事で数年 Go のコードを見ていますが、宣言された ok を使った if 分岐が comma ok idiom で書かれないのが珍しいなと思ったのでコメントした次第なのですが、強いて言うなら、自分が面接官をする場合、上記の idiom で書かれていたら「この人はある程度 Go で開発しているんだろうな」いう印象を持つかなとは思いました。
逆に Go に習熟していることをアピールしている求職者がこのような場合に comma ok idiom を使ってなかったら、ちょっと気になって「これあえてこうしてるんですか？」とか聞くかも知れません。

Reply:
☝️すいません、一部単語の使い方が誤っていたので修正しました。
（元のコード含めこのスレッドに関連する修正案や実装は comma ok idiom を用いたものでした）

Reply:
例えば下記のようにK8sのコードを見ると`if !ok`というのは数多く使われています。確かに`; ok`という書き方は良く見るとは思いますが、`if !ok`のような書き方が珍しいかというとそうではないと個人的には思います（あくまでも僕の狭い観測範囲での話ですが、、）。

https://cs.k8s.io/?q=if%20!ok&i=nope&files=&excludeFiles=&repos=

Reply:
参考まで、 Effective Go にも comma ok idiom を用いた判定を行う例として挙げられているものは、一つを除き全て `; ok` を使っていました。
https://go.dev/doc/effective_go

Reply:
> 例えば下記のようにK8sのコードを見るとif !okというのは数多く使われています。確かに; okという書き方は良く見るとは思いますが、if !okのような書き方が珍しいかというとそうではないと個人的には思います（あくまでも僕の狭い観測範囲での話ですが、、）。
> 
> https://cs.k8s.io/?q=if%20!ok&i=nope&files=&excludeFiles=&repos=

ありがとうございます。確かにたくさん記載箇所がありますね。僕も数年程度業務アプリケーションの開発をしているだけなので、 OSS のコードの実装は非常に参考になります。

Reply:
すいません今スマホで見てるので上記のページをちゃんと使えないのですが、 `if ok` の場合はどうでしょうか？

Reply:
ok が false であることのチェックを `; !ok` と書くのは見た記憶がないのですが（Effective Go にもなさそうでした）、 true であることのチェックは、 `; ok` と書かず別の行で実施するのは結構珍しいのかなと思っています。

Reply:
確かに検索が少し重いですよねw 調べたところ個数は出てこなかったんですが、`if !ok`が27248ファイル、`if ok`が22485個、`; ok`が31096ファイルでした。意外と`if ok`も使われてそうですね

Reply:
古くて安い Pixel 使ってるのでスペック低くてすいません😂
なるほどありがとうございます。こうやってある程度定量的な観測結果を見てみると学びが多いですね。勉強になりました！

Reply:
@皆さん
もし普段 Go を使っている方がいらっしゃいましたら、参考までにこちらの件ご意見貰えると嬉しいです！

Reply:
横からすみません。この2つの書き方にはスコープの観点で明確な違いがありそうなので、書き分けるのであればそこを意識するのがよいのでは、と思いました。

上記で comma ok idiom と言われている以下の書き方の場合、openBracket と ok のスコープが if 文内に閉じている認識です。この場合、スコープが狭いため意図せぬバグは防げる反面、その後のコードでは再利用できなくなります。

```
if openBracket, ok := closeToOpens[bracket]; ok {
```

OSS のコードの中でスコープを意識した使い分けがされているしているのかどうかまでは確認していませんが、1つ判断の基準になるのではと思ったのでコメントさせてもらいました。

Reply:
確かにそうですね。なので小さいスコープ内でしか使わないのであればなるべく`; ok`を使うようにするのが良いと思います。一方である程度のスコープを引き回して使いたい時は`if ok`のようにするのかなと思います

Reply:
同じく横から失礼します。私も気になってフレームワークのechoのソースコードを調べてみたところ、以下のような結果になりました。
`if !ok`: 8箇所
`if ok`: 3箇所
`; ok`: 26箇所

全箇所確認したわけではないですが、`if !ok`と`if ok`のパターンはthonda28さんのおっしゃる通りスコープの観点からそうしており、可読性に配慮して行を変えているわけではなかったです。
例えばこのようなコードがありました。
```Go
values, ok := c.Request().Form[sourceParam]
if !ok {
	return nil
}
return values
```

Reply:
ありがとうございます。スコープが広くなると、`; ok`とするとif節の行数が大きくなる感じになると思うので、`if ok`とするのは可読性にも貢献していることになると私は思います。

Reply:
例えば今回の場合、`if !ok`とすることでif節のネストを回避できていますが、`; ok`を使うとif節のスコープが広がり、結果的にネストが生まれてしまったりとコードの複雑性が高くなる結果につながっています。

Reply:
あとは`if !ok`を使うことで早期リターンのような形を取れるというのもあると思います。上記の提示してくださったコードもそういう意図があってやっているのかなと思いましたし、今回の私のコードも早期リターンみたいなことになっているため、それも認知負荷を下げることにつながっているのかなと思っています。

Reply:
はい、ネストについて同感です。あとif文が2行連続になることも気になりました。
上記の数字でも示したように`if !ok`はそれなりの頻度で使われていそうですし、使うことによってコードがより簡潔になるなら使ったほうが良さそうですね

そして自分なんか変なこと言ってましたね
> スコープの観点からそうしており、可読性に配慮して行を変えているわけではなかったです。

と書いてしまいましたが、
```Go
if values, ok := c.Request().Form[sourceParam]; !ok {
	return nil
} else {
        return values
}
```
とするより可読性が上がるから`if !ok`を使ってるんですよね。そしてスコープとかこの例ではあまり関係ない、、

Reply:
@seal-azarashi 話が長くなってしまって申し訳ないのですが、ちなみにEffective Goにcomma ok idiomを使うようにと明示的に言っている箇所ってありますでしょうか？Effective Goを見てみたのですが、comma ok idiomの名前は紹介されている箇所はあったのですが、特にそれ以外の言及をしている箇所を見つけられなかったので、、

> For obvious reasons this is called the “comma ok” idiom. In this example, if tz is present, seconds will be set appropriately and ok will be true; if not, seconds will be set to zero and ok will be false. Here's a function that puts it together with a nice error report:
https://go.dev/doc/effective_go

Reply:
あと、「comma ok idiom」と検索してみて上位に出てきた記事を数件見てみたのですが、いずれもcomma ok idiomを値を受け取ると同時にエラーハンドリングもできる便利な書き方として紹介されているのみで、日本のGoコミュニティにおいても`; !ok`, `if ok`, `if !ok`という書き方を排除するような風潮や慣習みたいなのはないのではないかと思ったのですが、どう思われますか

https://c-taquna.hatenablog.com/entry/2019/06/17/194407
https://note.com/ym202110/n/n5fe90360fad7
https://dev.to/saurabh975/comma-ok-in-go-l4f
https://ogataso.hatenablog.com/entry/2024/07/08/225459

Reply:
> Effective Goにcomma ok idiomを使うようにと明示的に言っている箇所ってありますでしょうか？

いえ、特に僕が見た限りはそのような記述はなかったと思います。

> ; okを使うように明示的に推奨している箇所が見当たらないので、; okのコード例が多いのも単にそういうコードが多いだけな気がしているんですがどうなんでしょうか、、

> あと、「comma ok idiom」と検索してみて上位に出てきた記事を数件見てみたのですが、いずれもcomma ok idiomを値を受け取ると同時にエラーハンドリングもできる便利な書き方として紹介されているのみで、日本のGoコミュニティにおいても; !ok, if ok, if !okという書き方を排除するような風潮や慣習みたいなのは一般的ではないのではないかと思ったのですが、どう思われますか

こちらのご認識のとおりだと思います。

Reply:
すいません書き方悪くて不安にさせてしまったのかなと反省しつつあるのですが、今回のコメントの内容としては、 comma ok idiom の ok が true であることをチェックする際に `; ok` を記述するのをよく見かけるというだけで、それ以上のこと、例えば `ok` を別の行で書くべきでないだとか、そういったことは特に考えていません。
面接で聞くかもとか書いてもいますが、特にその返答内容がクリティカルに合否に響くといったことも基本的にはないと考えてまして、あくまで雑談程度に聞くようなイメージでした。

変に推測する余地があるようなコメントをしてしまい失礼しました🙏

Reply:
ありがとうございます、そういうことだったのですね、承知いたしました🙏